File: CloneGraph.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    public class CloneGraph
    {
        //bfs
        /// Time Complexity : O(V+E) 
        // Space Complexity :O(V) 
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  No 
        public Node CloneGraph_BFS(Node node)
        {
            if (node == null) return node;
            Dictionary<Node, Node> map = new Dictionary<Node, Node>();
            Queue<Node> q = new Queue<Node>();
            Node newNode = new Node(node.val);
            map.Add(node, newNode);
            q.Enqueue(node);
            while (q.Count != 0)
            {
                Node curr = q.Dequeue();
                foreach (Node n in curr.neighbors)
                {
                    if (!map.ContainsKey(n))
                    {
                        map.Add(n, new Node(n.val));
                        q.Enqueue(n);
                    }
                    map.GetValueOrDefault(curr).neighbors.Add(map.GetValueOrDefault(n));
                }
            }
            return map.GetValueOrDefault(node);
        }


        Dictionary<Node, Node> map;
        public Node CloneGraph_DFS(Node node)
        {
            if (node == null) return node;
            map = new Dictionary<Node, Node>();
            Node newNode = new Node(node.val);
            map.Add(node, newNode);
            dfs(node);
            return map.GetValueOrDefault(node);
        }

        private void dfs(Node curr)
        {
            //base
            if (curr == null) return;
            //logic
            foreach (Node n in curr.neighbors)
            {
                if (!map.ContainsKey(n))
                {
                    map.Add(n, new Node(n.val));
                    dfs(n);
                }
                map.GetValueOrDefault(curr).neighbors.Add(map.GetValueOrDefault(n));
            }
        }
    }
}

File: clonegraph.swift
// Time Complexity : 0(n)
// Space Complexity : 0(n)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

    var hashmap = [Node:Node]()
    func cloneGraph(_ node: Node?) -> Node? {
        if node == nil {
            return nil
        }
        return dfs(node) 
    }
    func dfs(_ node:Node?) -> Node? {
        guard let node = node else { return nil}
        var newNode : Node? = nil
        if let nnode = hashmap[node] {
                newNode =  nnode
                return newNode
            } else {
                newNode =  Node(node.val) 
                hashmap[node] = newNode
            }
        for oldNode in node.neighbors {
            newNode?.neighbors.append(dfs(oldNode))
        }
        return newNode
    }